src/Compilers/CSharp/Portable/Lowering/NullCheckRewriter.cs

```diff
@@ -0,0 +1,97 @@
+    {
+        private readonly MethodSymbol _method;
+        private readonly SyntheticBoundNodeFactory _fact;
+        private NullCheckRewriter(
+            MethodSymbol method,
+            SyntaxNode syntax,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
+            _method = method;
+            _fact = new SyntheticBoundNodeFactory(method, syntax, compilationState, diagnostics);
+        }
+
+        internal static BoundStatement Rewrite(
+            BoundStatement body,
+            MethodSymbol method,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
+            if (!method.Parameters.Any(x => x is SourceParameterSymbolBase param
+                                            && param.IsNullChecked))
+            {
+                return body;
+            }
+
+            var rewriter = new NullCheckRewriter(method, body.Syntax, compilationState, diagnostics);
+            return (BoundStatement)rewriter.Visit(body);
+        }
+
+        public override BoundNode VisitBlock(BoundBlock node)
+        {
+            return AddNullChecksToBody(node);
+        }
+
+        private BoundNode AddNullChecksToBody(BoundBlock body)
+        {
+            var statementList = ArrayBuilder<BoundStatement>.GetInstance();
+            foreach (ParameterSymbol x in _method.Parameters)
+            {
+                if (x is SourceParameterSymbolBase param
+                    && param.IsNullChecked)
+                {
+                    if (param.Type.IsValueType && !param.Type.IsNullableTypeOrTypeParameter())
+                    {
+                        // PROTOTYPE : Warning or Error, see CodeGenNullCheckedParameterTests.TestNullCheckedSubstitution2
+                        continue;
+                    }
+                    var constructedIf = ConstructIfStatementForParameter(body, param);
+                    statementList.Add(constructedIf);
+                }
+            }
+            statementList.AddRange(body.Statements);
+
+            return _fact.Block(body.Locals, statementList.ToImmutableAndFree());
+        }
+
+        private BoundStatement ConstructIfStatementForParameter(BoundBlock body, SourceParameterSymbolBase parameter)
+        {
+            BoundExpression paramIsNullCondition;
+            var loweredLeft = _fact.Parameter(parameter);
+            if (loweredLeft.Type.IsNullableType())
+            {
+                var getNullableMethod = LocalRewriter.UnsafeGetNullableMethod(
+                                                    body.Syntax,
+                                                    loweredLeft.Type,
+                                                    SpecialMember.System_Nullable_T_get_HasValue,
+                                                    _method.DeclaringCompilation,
+                                                    _fact.Diagnostics);
```

Comment:
Are these arguments necessary? It looks like there is an overload with only the first three arguments that does the equivalent.

Reply:
It's not static - would you prefer me make the overload internal static?

Reply:
Ah, I missed that. In that case, calling this overload makes sense.

---
In reply to: [296337144](https://github.com/dotnet/roslyn/pull/36471#discussion_r296337144) [](ancestors = 296337144)